From 460c5bfa8ec3407a90b6ae377f49d65f122a09cc Mon Sep 17 00:00:00 2001
From: jason <jason@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Tue, 9 Sep 2014 11:57:25 +0000
Subject: [PATCH] 	PR c++/61214 	PR c++/62224 gcc/ 	*
 gimple-fold.c (can_refer_decl_in_current_unit_p): Don't allow 	reference to a
 DECL_EXTERNAL COMDAT. gcc/cp/ 	* decl2.c (decl_needed_p): Revert virtual
 functions change.

git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_9-branch@215061 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/gimple-fold.c                    |  3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)
 create mode 100644 gcc/testsuite/g++.dg/ipa/devirt-40.C

diff --git a/gcc/gimple-fold.c b/gcc/gimple-fold.c
index c6aea65..2527d29 100644
--- a/gcc/gimple-fold.c
+++ b/gcc/gimple-fold.c
@@ -146,7 +146,8 @@ can_refer_decl_in_current_unit_p (tree decl, tree from_decl)
          The second is important when devirtualization happens during final
          compilation stage when making a new reference no longer makes callee
          to be compiled.  */
-      if (!node || !node->definition || node->global.inlined_to)
+      if (!node || !node->definition
+	  || DECL_EXTERNAL (decl) || node->global.inlined_to)
 	{
 	  gcc_checking_assert (!TREE_ASM_WRITTEN (decl));
 	  return false;
-- 
2.1.2

